avocado/plugins/assets.py

```diff
@@ -203,6 +204,62 @@ def fetch_assets(test_file):
     return success, fail
 
 
+def fetch_command(references, verbose=True):
+    """
+    Iterate over the instrumented tests and try to fetch assets defined on it.
+    The objective of this function is to handle log messages, when needed.
+    :param references: Instrumented tests to be evaluated
+    :type references: list
+    :param verbose: Whether log messages should be displayed
+    :returns: exit code from assets fetch execution
+    """
+    exitcode = exit_codes.AVOCADO_ALL_OK
+    for test_file in references:
+        if os.path.isfile(test_file) and test_file.endswith('.py'):
+            if verbose:
+                LOG_UI.info('Fetching assets from %s.', test_file)
+            success, fail = fetch_assets(test_file)
+
+            if verbose:
+                for asset_file in success:
+                    LOG_UI.info('  File %s fetched or already on'
+                                ' cache.', asset_file)
+                for asset_file in fail:
+                    LOG_UI.error(asset_file)
+
+            if fail:
+                exitcode |= exit_codes.AVOCADO_FAIL
+        else:
+            if verbose:
+                LOG_UI.warning('No such file or file not supported: %s',
+                               test_file)
+            exitcode |= exit_codes.AVOCADO_FAIL
+
+    return exitcode
+
+
+class FetchAssetJob(JobPreTests):  # pylint: disable=R0903
+    """
+    Implements the assets fetch job pre tests. This has the same effect of
+    running the 'avocado assets fetch INSTRUMENTED', but it runs during the
+    test execution, before the actual test starts.
+    """
+    name = "fetchasset"
+    description = "Fetch assets before the test run"
+
+    def __init__(self, config=None):
+        self.config = config
+
+    def pre_tests(self, job):
+        exitcode = fetch_command(self.config['references'], verbose=False)
```

Comment:
So, instead of `self.config['reference']` you *could* use `job.config.get('references', [])`, but there's a catch.  One of the goals of using a `JobPreTests` plugin, is that, not only you'd have access to the references, but you have access to the test suite (think of it as the resolved, filtered references).

For instance, when running:

```
$ avocado list tests/acceptance/boot_linux_console.py 
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_x86_64_pc
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_mips_malta
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_mips64el_malta
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_mips_malta_cpio
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_mips64el_malta_5KEc_cpio
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_mips_malta32el_nanomips_4k
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_mips_malta32el_nanomips_16k_up
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_mips_malta32el_nanomips_64k_dbg
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_aarch64_virt
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_arm_virt
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_arm_emcraft_sf2
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_arm_raspi2_uart0
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_arm_exynos4210_initrd
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_s390x_s390_ccw_virtio
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_alpha_clipper
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_ppc64_pseries
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_m68k_q800
````

And consequently:

```
$ avocado assets fetch tests/acceptance/boot_linux_console.py 
Fetching assets from tests/acceptance/boot_linux_console.py.
  File https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/29/Everything/x86_64/os/images/pxeboot/vmlinuz fetched or already on cache.
  File http://snapshot.debian.org/archive/debian/20130217T032700Z/pool/main/l/linux-2.6/linux-image-2.6.32-5-4kc-malta_2.6.32-48_mips.deb fetched or already on cache.
  File http://snapshot.debian.org/archive/debian/20130217T032700Z/pool/main/l/linux-2.6/linux-image-2.6.32-5-5kc-malta_2.6.32-48_mipsel.deb fetched or already on cache.
  File http://snapshot.debian.org/archive/debian/20160601T041800Z/pool/main/l/linux/linux-image-4.5.0-2-4kc-malta_4.5.5-1_mips.deb fetched or already on cache.
  File https://github.com/groeck/linux-build-test/raw/8584a59ed9e5eb5ee7ca91f6d74bbb06619205b8/rootfs/mips/rootfs.cpio.gz fetched or already on cache.
  File https://github.com/philmd/qemu-testing-blob/raw/9ad2df38/mips/malta/mips64el/vmlinux-3.19.3.mtoman.20150408 fetched or already on cache.
  File https://github.com/groeck/linux-build-test/raw/8584a59e/rootfs/mipsel64/rootfs.mipsel64r1.cpio.gz fetched or already on cache.
  File https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/29/Everything/aarch64/os/images/pxeboot/vmlinuz fetched or already on cache.
  File https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/29/Everything/armhfp/os/images/pxeboot/vmlinuz fetched or already on cache.
  File https://raw.githubusercontent.com/Subbaraya-Sundeep/qemu-test-binaries/fa030bd77a014a0b8e360d3b7011df89283a2f0b/u-boot fetched or already on cache.
  File https://raw.githubusercontent.com/Subbaraya-Sundeep/qemu-test-binaries/fa030bd77a014a0b8e360d3b7011df89283a2f0b/spi.bin fetched or already on cache.
  File http://archive.raspberrypi.org/debian/pool/main/r/raspberrypi-firmware/raspberrypi-kernel_1.20190215-1_armhf.deb fetched or already on cache.
  File https://snapshot.debian.org/archive/debian/20190928T224601Z/pool/main/l/linux/linux-image-4.19.0-6-armmp_4.19.67-2+deb10u1_armhf.deb fetched or already on cache.
  File https://github.com/groeck/linux-build-test/raw/2eb0a73b5d5a28df3170c546ddaaa9757e1e0848/rootfs/arm/rootfs-armv5.cpio.gz fetched or already on cache.
  File https://archives.fedoraproject.org/pub/archive/fedora-secondary/releases/29/Everything/s390x/os/images/kernel.img fetched or already on cache.
  File http://archive.debian.org/debian/dists/lenny/main/installer-alpha/current/images/cdrom/vmlinuz fetched or already on cache.
  File https://archives.fedoraproject.org/pub/archive/fedora-secondary/releases/29/Everything/ppc64le/os/ppc/ppc64/vmlinuz fetched or already on cache.
  File http://ftp.ports.debian.org/debian-ports/pool-m68k/main/l/linux/kernel-image-5.3.0-1-m68k-di_5.3.7-1_m68k.udeb fetched or already on cache.
```

But, many different tests may live in a single file, and suppose one runs (which is similarly to what is executed when one only builds a given QEMU, say x86_64, target on the acceptance tests):

```
$ avocado list -t arch:x86_64 tests/acceptance/boot_linux_console.py 
INSTRUMENTED tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_x86_64_pc
```

So here all but one asset (`https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/29/Everything/x86_64/os/images/pxeboot/vmlinuz`) needs to be fetched... there's no need to fetch any asset for `tests/acceptance/boot_linux_console.py`.

The way to go is to look at the `job.test_suite` attribute.  Here I have:

```
[('BootLinuxConsole', {'name': 'tests/acceptance/boot_linux_console.py:BootLinuxConsole.test_x86_64_pc', 'modulePath': 'tests/acceptance/boot_linux_console.py', 'methodName': 'test_x86_64_pc', 'tags': {'arch': {'x86_64'}, 'machine': {'pc'}}})]
```

https://github.com/clebergnu/avocado/blob/mini_rfc_job_phases/docs/rfcs/job_phases.rst should give an idea of what is available at a given "job phase".  Here, you could look at `job.test_suite` and from it, only look for the assets for the test methods in the suite.

Another tip is that if the first member of each tuple is a string, it means that it's an Avocado INSTRUMENTED tests (that will be really loaded and executed by the runner).